qiskit/transpiler/passes/layout/vf2_post_layout.py

```diff
@@ -0,0 +1,311 @@
+            if len_args == 2:
+                if node.qargs[0] not in im_graph_node_map:
+                    im_graph_node_map[node.qargs[0]] = im_graph.add_node(defaultdict(int))
+                    reverse_im_graph_node_map[im_graph_node_map[node.qargs[0]]] = node.qargs[0]
+                if node.qargs[1] not in im_graph_node_map:
+                    im_graph_node_map[node.qargs[1]] = im_graph.add_node(defaultdict(int))
+                    reverse_im_graph_node_map[im_graph_node_map[node.qargs[1]]] = node.qargs[1]
+                edge = (im_graph_node_map[node.qargs[0]], im_graph_node_map[node.qargs[1]])
+                if im_graph.has_edge(*edge):
+                    im_graph.get_edge_data(*edge)[node.name] += 1
+                else:
+                    weight = defaultdict(int)
+                    weight[node.name] += 1
+                    im_graph.add_edge(*edge, weight)
+            if len_args >= 3:
+                self.property_set[
+                    "VF2PostLayout_stop_reason"
+                ] = VF2PostLayoutStopReason.MORE_THAN_2Q
+                return
+
+        if self.target is not None:
+            cm_graph = PyDiGraph(multigraph=False)
+            cm_graph.add_nodes_from(
+                [self.target.operation_names_for_qargs((i,)) for i in range(self.target.num_qubits)]
+            )
+            for qargs in self.target.qargs:
+                len_args = len(qargs)
+                # If qargs == 1 we already populated it and if qargs > 2 there are no instructions
+                # using those in the circuit because we'd have already returned by this point
+                if len_args == 2:
+                    cm_graph.add_edge(
+                        qargs[0], qargs[1], self.target.operation_names_for_qargs(qargs)
+                    )
+            cm_nodes = list(cm_graph.node_indexes())
+        else:
+            cm_graph = self.coupling_map.graph
+            cm_nodes = list(cm_graph.node_indexes())
+            if self.seed != -1:
+                random.Random(self.seed).shuffle(cm_nodes)
+                shuffled_cm_graph = type(cm_graph)()
+                shuffled_cm_graph.add_nodes_from(cm_nodes)
+                new_edges = [
+                    (cm_nodes[edge[0]], cm_nodes[edge[1]]) for edge in cm_graph.edge_list()
+                ]
+                shuffled_cm_graph.add_edges_from_no_data(new_edges)
+                cm_nodes = [k for k, v in sorted(enumerate(cm_nodes), key=lambda item: item[1])]
+                cm_graph = shuffled_cm_graph
+
+        logger.debug("Running VF2 to find post transpile mappings")
+        if self.target:
+            mappings = vf2_mapping(
+                cm_graph,
+                im_graph,
+                node_matcher=_target_match,
+                edge_matcher=_target_match,
+                subgraph=True,
+                id_order=False,
+                induced=False,
+                call_limit=self.call_limit,
+            )
```

Comment:
does this return ALL subgraphs and then we score them here? this should be pretty expensive right?

Also say we have a line of 5 qubits and we have chose subset 10 to 1., there are two possibilities based on the orientation of layout. It could be either `10-11-12-13-14` or `14-13-12-11-10`. Does it return them both and score separately, or just one? For say a ring of 12 on heavy-hex, there would be 12 orientations.

Reply:
It does `mappings` here is an iterator and each step will compute a new isomorphic mapping. It can be slow and that is why we set `call_limit` which is the number of internal state visits the function will try so we limit the amount of time we try waiting. In the preset passmanagers we set this at increasing large number so that we spend at most ~100ms for level1, ~10sec for level 2, and ~60 sec for level 3.

As for the orientation it will try both because it is a directed graph and we're using strict edges. This is actually necessary especially for the backendv2/target path because in that path we might not have all gates available in both directions (which is what the matcher functions here are checking). Also the scores can be different because we'd potentially end up with different gate counts on each qubit and 2q link which would change the score between each orientation.

Reply:
oh i see, it's looking for directed subgraphs. I think this can be problematic. Suppose I have a circuit that has been laid out on qubits `0 ---> 1 ---> 2` in this direction. So the circuit would be:
```
----*------
    |
----X--*---
       |
-------X---
```

Now if these qubits are very noisy but qubits `3 --> 4 <--- 5` are very good, they will not be chosen because they don't have the right direction. But actually fixing direction is trivial. I suggest that choosing layouts should be on undirected graphs (only find good subsets). Then apply a post-post-layout direction fixing pass.

For the case I brought up which is when there are multiple orientations (2 for a line, 12 for a ring of 12), I think this can be an interesting follow-up of choosing the best among those. But already choosing the best subset among many will go a long way.

Reply:
Yeah I considered using undirected edges and using gate direction when I wrote this, the problem with it is to do an undirected subgraph search we'd basically need to rerun the transpiler after the routing phase again. We can easily do `GateDirection` after applying the layout to fix this, but then we've potentially injected a bunch of hadamards into the circuit so we need the basis translator to convert that to the native basis set. This will then require the optimization passes to run because the basis translator output can likely be simplified. We basically end up rerunning most of the transpiler at that point. Typing all of this now though has made met think of a potentially interesting follow on we can play with where we could add a strict direction flag to this pass and then add `VF2PostLayout` and `ApplyLayout` to end of the optimization loop with that flag set `False`.

The way I was viewing this pass was given the hard constraints on the backend can we find a better qubit selection with lower noise and if not we don't do anything. So we do miss the opportunity for `3 -> 4 < - 5` if there are no compatible 2q gates on that direction but it is just a heuristic and that seemed ok . Especially with `BackendV2` where the gates are defined per qubit (like in your example if `0 -> 1 -> 2` was all in cx but `3 -> 4 <- 5` was only ecr). This seemed the better path to start since in all my tests it was able to find better layouts. At least for all the current backends with connectivity constraints this won't come up since they all currently define bidirectional edges with the same error rates.

Reply:
> We basically end up rerunning most of the transpiler at that point.

If we order the passes right I don't think there will be a duplication. In my mind the order should be something like this:

1) high-level synthesis (toffolis, cliffords, unitaries, etc.) to reduce the circuit to 1 & 2 qubit gates
2) layout + routing + post-layout
3) other parts including 1-& 2-qubit synthesis, optimization, scheduling. Since these just make local changes on 1 or 2 qubits at a time, they don't alter the mapping.

So I was thinking that this `PostLayout` pass can be done in stage 2 (basically to improve the layout). But if the scoring mechanism relies on the gates exactly being in the Target, then it wouldn't work. 
For it to work we would need a more relaxed scoring that can approximate. e.g. if there's a 2-qubit unitary it can assign a score to it based on looking at the 2-qubit error rates on that link. It wouldn't be exact, but I think it would be good enough. Since the whole scoring is approximate anyway. (I think soon the devices will report the native cx direction only btw)

Reply:
I'll give it a try, I'm curious to see the difference between running it at different spots in the pass manager. I'm also wondering if there value in doing post layout > 1 time in a pipeline, like if we did it with looser constraints at the end of 2 and after 3 with the stricter constraints.

FWIW, I did this after 3 because I thought it would be better because we have the complete circuit so we can see how many gates get run on each of the qubits and get the full error rates with each layout. Especially since `DenseLayout` is already noise aware so it should be picking similar qubits already. But it's definitely worth testing and checking to see what makes a bigger impact on result quality.

Reply:
I've been testing this locally running on hardware and most of the time the layouts are the same. But there has been 1 time so far where the layouts were significantly different and the undirected case doing it right after routing was significantly better. So I'm going to adjust the preset pass managers to do it this way in the PR.

Reply:
I added the undirected mode in https://github.com/Qiskit/qiskit-terra/pull/7862/commits/3b3c0a05532c86d0ec340724422538d180646d6d and started using that in the preset pass managers after routing in https://github.com/Qiskit/qiskit-terra/pull/7862/commits/5e90148f04e1a3c513473470647fa8e4648f45db